### ISSUE TITLE
output true CER for checkpoints (at least the final one)

### ISSUE THREAD

**Issue by bertsky** [tesseract-ocr/tesstrain#261]
AFAICS, `lstmtraining` produces two types of figures for measuring the error:
1. **bag-of-character training error** (on `list.train`): this is shown as
   - `char train=%.3f%%` every 100 iterations
   - `Finished! Error rate = %.2f` at the end
   - first figure in the file name of the checkpoint files
2. **bag-of-character testing error** (on `list.eval`): this is shown as
   - `Eval Char error rate=%f` (but also in percent) whenever (IIUC) training error reached a new minimum

<details><summary>Call-tree for 1:</summary>
<p>

- https://github.com/tesseract-ocr/tesseract/blob/7a308edcb1fc7455008b531bc2a49de583d7b171/src/training/unicharset/lstmtrainer.cpp#L945-L947
- https://github.com/tesseract-ocr/tesseract/blob/7a308edcb1fc7455008b531bc2a49de583d7b171/src/training/unicharset/lstmtrainer.cpp#L1261-L1286
- https://github.com/tesseract-ocr/tesseract/blob/7a308edcb1fc7455008b531bc2a49de583d7b171/src/training/unicharset/lstmtrainer.cpp#L1201-L1221
- https://github.com/tesseract-ocr/tesseract/blob/7a308edcb1fc7455008b531bc2a49de583d7b171/src/training/unicharset/lstmtrainer.cpp#L1325-L1339
- https://github.com/tesseract-ocr/tesseract/blob/7a308edcb1fc7455008b531bc2a49de583d7b171/src/training/unicharset/lstmtrainer.cpp#L369-L378

</p></details>
<details><summary>Call-tree for 2:</summary>
<p>

- https://github.com/tesseract-ocr/tesseract/blob/7a308edcb1fc7455008b531bc2a49de583d7b171/src/training/unicharset/lstmtester.cpp#L48-L134
- https://github.com/tesseract-ocr/tesseract/blob/7a308edcb1fc7455008b531bc2a49de583d7b171/src/training/unicharset/lstmtrainer.h#L151-L154
- https://github.com/tesseract-ocr/tesseract/blob/7a308edcb1fc7455008b531bc2a49de583d7b171/src/training/lstmtraining.cpp#L188-L197
- https://github.com/tesseract-ocr/tesseract/blob/7a308edcb1fc7455008b531bc2a49de583d7b171/src/training/lstmtraining.cpp#L217

</p></details>

However, IMHO this is _higly misleading_, for two reasons:
- when you pass a validation set, you expect checkpoint and final error rate to refer to that – not the training error
- today you expect `error` to mean character error rate (CER) – not the bag-of-character error (average character count confusion per line)

Or am I missing something very large and obvious?

(I do usually get an order of magnitude larger error rate when measuring via prediction and Levenshtein distance compared to lstmtraining. So at least there's empirical evidence something is not right with lstmtraining's figures. I can provide them when needed here.)

**Comment by bertsky** [comment]
(Issue probably needs to be transferred to the main repo.)

**Comment by stweil** [comment]
That's a problem which might have been less important before using tesstrain: the initial training used lines with artificial ground truth. As those lines were long (all with similar length), CER and bag-of-character error would not differ much.

**Comment by bertsky** [comment]
> That's a problem which might have been less important before using tesstrain: the initial training used lines with artificial ground truth. As those lines were long (with similar lengths), CER and bag-of-character error would not differ.
> 
> Both numbers will only differ much if the CER depends on the line length. Calculating a weighted mean error might be sufficient to get a CER.

It's true bag-of-character error rate (BCER) gets more "stable" with line length, and in that sense (only) CER and BCER tend to be more similar on longer lines. But BCER also always systematically underestimates CER, and both measures _also_ diverge (get more and more uncorrelated) the larger the error is in total. (Thus, no weighting scheme for shorter lines can make up for this bias.)

And as I said, my empirical data suggest that there's a huge difference. (But I'll still have to compute CER measurements of the _training_ error for a direct comparison.)

BTW, `lstmeval` is no better than `lstmtraining` in this regard, as it delegates to the same error metric.

Now, CER vs. BCER is quite a blow, but it's (relatively) easy to fix: just incorporate _some_ C++ implementation of global sequence alignment (Needleman-Wunsch algorithm) and distance measure (Damerau-Levenshtein metric) into Tesseract and use that for the evaluation data in a checkpoint.

The other problem mentioned above, i.e. training error instead of test error, is even bigger I believe: Even the checkpointing itself merely runs on training error minima (since `MaintainCheckpoints` runs on the trainer instance and merely calls the tester instance for log messages). So we simply _don't know_ which checkpoints (or even which iterations) performed best on the test set. I'm afraid this means that all models must be retrained after implementing proper test error checkpointing / early-stopping.

**Comment by bertsky** [comment]
> (Issue probably needs to be transferred to the main repo.)

@stweil please transfer!

**Comment by stweil** [comment]
Should I move this issue to the `tesseract` repository?

**Comment by bertsky** [comment]
> Should I move this issue to the `tesseract` repository?

Please do!
